scipy/special/specfun.py

```diff
@@ -5,31 +5,11 @@
 from scipy._lib.deprecation import _sub_module_deprecation
 
 __all__ = [  # noqa: F822
```

Comment:
maybe we just do, not sure why this works and the other doesn't
```suggestion
# ruff: noqa: F822
__all__ = [
```

Reply:
Linter looks happy now. Do we still need the `noqa: F822`? Can we try removing it now that we have the `from` imports above?

Reply:
Not sure why this specific instance was failing we, use this pattern in loads of other places without issue!

Reply:
So the linter seems happy now, but I'm getting circular import errors with test_public_api that I've never seen before. https://github.com/scipy/scipy/pull/20320#discussion_r1558233772

All about specfun.py, which is essentially just an empty file. Can I just delete those tests, @j-bowhay?

Reply:
Which tests sorry? On an unrelated note I am having difficulty building relating to mdspan to work out what's going on
<details open>
<summary>details</summary>
<be>

```
💻  ninja -C /home/jakeb/development/scipy/build -j8
ninja: Entering directory `/home/jakeb/development/scipy/build'
[12/1356] Compiling C++ object scipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p/_gufuncs.cpp.o
FAILED: scipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p/_gufuncs.cpp.o 
/home/jakeb/miniconda3/envs/scipy-dev/bin/x86_64-conda-linux-gnu-c++ -Iscipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p -Iscipy/special -I../scipy/special -Iscipy/_lib -I../scipy/_lib -I../scipy/_build_utils/src -I../../../miniconda3/envs/scipy-dev/lib/python3.11/site-packages/numpy/core/include -I/home/jakeb/miniconda3/envs/scipy-dev/include/python3.11 -fvisibility=hidden -fvisibility-inlines-hidden -fdiagnostics-color=always -D_GLIBCXX_ASSERTIONS=1 -D_FILE_OFFSET_BITS=64 -Wall -Winvalid-pch -std=c++17 -O2 -g -fvisibility-inlines-hidden -fmessage-length=0 -march=nocona -mtune=haswell -ftree-vectorize -fPIC -fstack-protector-strong -fno-plt -O2 -ffunction-sections -pipe -isystem /home/jakeb/miniconda3/envs/scipy-dev/include -DNDEBUG -D_FORTIFY_SOURCE=2 -O2 -isystem /home/jakeb/miniconda3/envs/scipy-dev/include -fPIC -DNPY_NO_DEPRECATED_API=NPY_1_9_API_VERSION -DSP_SPECFUN_ERROR -MD -MQ scipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p/_gufuncs.cpp.o -MF scipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p/_gufuncs.cpp.o.d -o scipy/special/_gufuncs.cpython-311-x86_64-linux-gnu.so.p/_gufuncs.cpp.o -c ../scipy/special/_gufuncs.cpp
In file included from ../scipy/special/ufunc.h:15,
                 from ../scipy/special/_gufuncs.cpp:3:
../scipy/special/special/mdspan.h: In instantiation of 'class std::extents<long int, 18446744073709551615>':
../scipy/special/special/mdspan.h:120:11:   required from 'class std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>'
../scipy/special/special/legendre.h:19:6:   required from 'void special::lpn(T, OutputVec1, OutputVec2) [with T = float; OutputVec1 = std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>; OutputVec2 = std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>]'
../scipy/special/_gufuncs.cpp:67:67:   required from here
../scipy/special/special/mdspan.h:40:43: error: 'std::extents<Index, Extents>::m_dexts' has incomplete type
   40 |     array<index_type, sizeof...(Extents)> m_dexts;
      |                                           ^~~~~~~
In file included from /home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/bits/unique_ptr.h:36,
                 from /home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/memory:75,
                 from ../scipy/special/ufunc.h:2:
/home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/tuple:1595:45: note: declaration of 'struct std::array<long int, 1>'
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
../scipy/special/special/mdspan.h: In instantiation of 'class std::layout_stride::mapping<std::extents<long int, 18446744073709551615> >':
../scipy/special/special/mdspan.h:128:18:   required from 'class std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>'
../scipy/special/special/legendre.h:19:6:   required from 'void special::lpn(T, OutputVec1, OutputVec2) [with T = float; OutputVec1 = std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>; OutputVec2 = std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>]'
../scipy/special/_gufuncs.cpp:67:67:   required from here
../scipy/special/special/mdspan.h:72:49: error: 'std::layout_stride::mapping<Extents>::m_strides' has incomplete type
   72 |         array<index_type, extents_type::rank()> m_strides;
      |                                                 ^~~~~~~~~
/home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/tuple:1595:45: note: declaration of 'struct std::array<long int, 1>'
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
../scipy/special/special/mdspan.h: In instantiation of 'class std::extents<long int, 18446744073709551615, 18446744073709551615>':
../scipy/special/special/mdspan.h:120:11:   required from 'class std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>'
../scipy/special/special/legendre.h:127:6:   required from 'void special::lpmn(T, long int, OutputMat1, OutputMat2) [with T = float; OutputMat1 = std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; OutputMat2 = std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>]'
../scipy/special/_gufuncs.cpp:73:69:   required from here
../scipy/special/special/mdspan.h:40:43: error: 'std::extents<Index, Extents>::m_dexts' has incomplete type
   40 |     array<index_type, sizeof...(Extents)> m_dexts;
      |                                           ^~~~~~~
/home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/tuple:1595:45: note: declaration of 'struct std::array<long int, 2>'
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
../scipy/special/special/mdspan.h: In instantiation of 'class std::layout_stride::mapping<std::extents<long int, 18446744073709551615, 18446744073709551615> >':
../scipy/special/special/mdspan.h:128:18:   required from 'class std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>'
../scipy/special/special/legendre.h:127:6:   required from 'void special::lpmn(T, long int, OutputMat1, OutputMat2) [with T = float; OutputMat1 = std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; OutputMat2 = std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>]'
../scipy/special/_gufuncs.cpp:73:69:   required from here
../scipy/special/special/mdspan.h:72:49: error: 'std::layout_stride::mapping<Extents>::m_strides' has incomplete type
   72 |         array<index_type, extents_type::rank()> m_strides;
      |                                                 ^~~~~~~~~
/home/jakeb/miniconda3/envs/scipy-dev/x86_64-conda-linux-gnu/include/c++/12.3.0/tuple:1595:45: note: declaration of 'struct std::array<long int, 2>'
 1595 |   template<typename _Tp, size_t _Nm> struct array;
      |                                             ^~~~~
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = float; Extents = std::extents<long int, 18446744073709551615>; AccessorPolicy = std::default_accessor<float>; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {float, std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >, std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >}; long unsigned int ...I = {0, 1, 2}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {float, std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >, std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >}]'
../scipy/special/_gufuncs.cpp:67:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 1> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 1> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<float, std::extents<long int, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = double; Extents = std::extents<long int, 18446744073709551615>; AccessorPolicy = std::default_accessor<double>; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {double, std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >, std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >}; long unsigned int ...I = {0, 1, 2}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {double, std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >, std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >}]'
../scipy/special/_gufuncs.cpp:67:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 1> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 1> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<double, std::extents<long int, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = std::complex<float>; Extents = std::extents<long int, 18446744073709551615>; AccessorPolicy = std::default_accessor<std::complex<float> >; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {std::complex<float>, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >}; long unsigned int ...I = {0, 1, 2}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {std::complex<float>, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >}]'
../scipy/special/_gufuncs.cpp:67:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 1> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 1> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = std::complex<double>; Extents = std::extents<long int, 18446744073709551615>; AccessorPolicy = std::default_accessor<std::complex<double> >; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {std::complex<double>, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >}; long unsigned int ...I = {0, 1, 2}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {std::complex<double>, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >}]'
../scipy/special/_gufuncs.cpp:67:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 1> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 1> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = float; Extents = std::extents<long int, 18446744073709551615, 18446744073709551615>; AccessorPolicy = std::default_accessor<float>; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {float, long int, std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >, std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >}; long unsigned int ...I = {0, 1, 2, 3}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {float, long int, std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >, std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<float> >}]'
../scipy/special/_gufuncs.cpp:73:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 2> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 2> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<float, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = double; Extents = std::extents<long int, 18446744073709551615, 18446744073709551615>; AccessorPolicy = std::default_accessor<double>; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {double, long int, std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >, std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >}; long unsigned int ...I = {0, 1, 2, 3}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {double, long int, std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >, std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<double> >}]'
../scipy/special/_gufuncs.cpp:73:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 2> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 2> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<double, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = std::complex<float>; Extents = std::extents<long int, 18446744073709551615, 18446744073709551615>; AccessorPolicy = std::default_accessor<std::complex<float> >; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {std::complex<float>, long int, long int, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >}; long unsigned int ...I = {0, 1, 2, 3, 4}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {std::complex<float>, long int, long int, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >, std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<float> > >}]'
../scipy/special/_gufuncs.cpp:78:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 2> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 2> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<std::complex<float>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&'
../scipy/special/ufunc.h: In instantiation of 'void from_pointer(char*, std::mdspan<T, Extents, std::layout_stride, AccessorPolicy>&, const npy_intp*, const npy_intp*) [with T = std::complex<double>; Extents = std::extents<long int, 18446744073709551615, 18446744073709551615>; AccessorPolicy = std::default_accessor<std::complex<double> >; npy_intp = long int]':
../scipy/special/ufunc.h:148:17:   required from 'T from_pointer(char*, const npy_intp*, const npy_intp*) [with T = std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>; npy_intp = long int]'
../scipy/special/ufunc.h:192:36:   required from 'static void ufunc_traits<void(Args ...), std::integer_sequence<long unsigned int, I ...> >::loop_func(char**, const npy_intp*, const npy_intp*, void*) [with Args = {std::complex<double>, long int, long int, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >}; long unsigned int ...I = {0, 1, 2, 3, 4}; npy_intp = long int]'
../scipy/special/ufunc.h:215:51:   required from 'SpecFun_Func::SpecFun_Func(Res (*)(Args ...)) [with Res = void; Args = {std::complex<double>, long int, long int, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >, std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride, std::default_accessor<std::complex<double> > >}]'
../scipy/special/_gufuncs.cpp:78:26:   required from here
../scipy/special/ufunc.h:126:44: error: 'std::array<long int, 2> strides' has incomplete type
  126 |     std::array<ptrdiff_t, Extents::rank()> strides;
      |                                            ^~~~~~~
../scipy/special/ufunc.h:131:44: error: 'std::array<long int, 2> exts' has incomplete type
  131 |     std::array<ptrdiff_t, Extents::rank()> exts;
      |                                            ^~~~
../scipy/special/ufunc.h:136:9: error: no match for 'operator=' (operand types are 'std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>' and '<brace-enclosed initializer list>')
  136 |     dst = {reinterpret_cast<T *>(src), {exts, strides}};
      |     ~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(const std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&)'
  112 | class mdspan {
      |       ^~~~~~
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'const std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&'
../scipy/special/special/mdspan.h:112:7: note: candidate: 'constexpr std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>& std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>::operator=(std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&)'
../scipy/special/special/mdspan.h:112:7: note:   no known conversion for argument 1 from '<brace-enclosed initializer list>' to 'std::mdspan<std::complex<double>, std::extents<long int, 18446744073709551615, 18446744073709551615>, std::layout_stride>&&'
[19/1356] Compiling C++ object scipy/special/_special_ufuncs.cpython-311-x86_64-linux-gnu.so.p/_special_ufuncs.cpp.o
ninja: build stopped: subcommand failed.
Build failed!
```
</details>

Reply:
Hmm, it's building on CI though...? In any case, I just took a guess at what your local issue might be, and added something.

Reply:
Ok great thanks that fixed my issue

Reply:
Ok the circular import is just a missing line from meson. I am convinced the original lint fail is a ruff bug. If I used flake8 there are no complaints. Anyway this has everything passing locally for me, I forgot originally where `pbdv` lives hence another change to private_modules
```
diff --git a/scipy/special/meson.build b/scipy/special/meson.build
index 1bd4afa1b..8663d694b 100644
--- a/scipy/special/meson.build
+++ b/scipy/special/meson.build
@@ -393,6 +393,7 @@ python_sources = [
   '_orthogonal.py',
   '_orthogonal.pyi',
   '_sf_error.py',
+  'specfun.py',
   '_spfun_stats.py',
   '_spherical_bessel.py',
   '_support_alternative_backends.py',
diff --git a/scipy/special/specfun.py b/scipy/special/specfun.py
index d5736e787..f67072cad 100644
--- a/scipy/special/specfun.py
+++ b/scipy/special/specfun.py
@@ -3,10 +3,9 @@
 # included below.
 
 from scipy._lib.deprecation import _sub_module_deprecation
-from ._basic import clpmn, lpmn, lpn, lqmn
-from ._specfun import pbdv
 
-__all__ = [  # noqa: F822
+ # ruff: noqa: F822
+__all__ = [
     'clpmn',
     'lpmn',
     'lpn',
@@ -21,5 +20,5 @@ def __dir__():
 
 def __getattr__(name):
     return _sub_module_deprecation(sub_package="special", module="specfun",
-                                   private_modules=["_basic"], all=__all__,
+                                   private_modules=["_basic", "_specfun"], all=__all__,
                                    attribute=name)
```

Reply:
Ok looks like things relating to this part at least are green

Reply:
Yep! Many thanks @j-bowhay, that was super helpful.

Reply:
Still not sure I understand why the CI didn't need the array include but anyway